### PR TITLE
Fix WebSocket key compare for some servers

### PR DESCRIPTION
**Pull request by ADiea**
Sometimes server returns lowercase or different case headers, [allowed by RFC](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2). Key will not match then. Happened on Heroku.

- compare key header lowercase for websocket
- small fix check TCP pointer causes reboot sometimes
- make Netutils::debugPrintTcpList static to call it without object instance

### CODE REVIEW

File: Sming/SmingCore/Network/TcpConnection.cpp [hunk 1]
```diff
@@ -347,7 +347,7 @@ void TcpConnection::closeTcpConnection(tcp_pcb *tpcb)
 
 void TcpConnection::flush()
 {
-	if (tcp->state == ESTABLISHED)
+	if (tcp && tcp->state == ESTABLISHED)
```
Comment (slaff) [inline Sming/SmingCore/Network/TcpConnection.cpp]: That is a change that needs to be moved to separate PR.

File: Sming/SmingCore/Network/NetUtils.cpp [hunk 2]
```diff
@@ -180,3 +180,35 @@ void NetUtils::debugPrintTcpList()
 	if(prt_none) debugf("LWIP_DEBUG: none\r\n");
 	return;
 }
+
+//search lowercase
+char* NetUtils::strstrl(char* source, const char* sequence)
```
Comment (slaff) [inline Sming/SmingCore/Network/NetUtils.cpp]: Why not just combine `strstr` and `tolower` ? You can see `Wiring/WString.cpp` for example usage.

File: Sming/SmingCore/Network/NetUtils.h [hunk 3]
```diff
@@ -27,11 +27,12 @@ class NetUtils
 	static int pbufFindStr(pbuf *buf, const char* wtf, int startPos = 0);
 	static char* pbufAllocateStrCopy(pbuf *buf, int startPos, int length);
 	static String pbufStrCopy(pbuf *buf, int startPos, int length);
+	static char* strstrl(char* source, const char* sequence);
```
Comment (slaff) [inline Sming/SmingCore/Network/NetUtils.h]: This method does not have place here.

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 4]
```diff
@@ -101,17 +103,35 @@ void WebsocketClient::onError(err_t err)
...
+	if(!endKey || endKey - buf > size)
+	{
+		debugf("wscli cannot find key(2) reason:%s", endKey ? "out of bounds":"NULL");
+		return false;
+	}
+	//terminate serverKey
+	*endKey = 0;
+	debugf("wscli search key %s", serverKey);
+
+
```
Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: Coding Style: One empty new line as a separator should be enough.

File: Sming/SmingCore/Network/NetUtils.h [hunk 5]
```diff
@@ -27,11 +27,12 @@ class NetUtils
...
 
 	static bool FixNetworkRouting();
 
 	// Debug
-	void debugPrintTcpList();
+	static void debugPrintTcpList();
```
Comment (slaff) [inline Sming/SmingCore/Network/NetUtils.h]: Is this change related to Websocket Key comparison? If not - please move it to another PR.

**Comment by ADiea** [comment]
Remade the PR, added indexOfIgnoreCase to String class

File: Sming/SmingCore/Network/TcpConnection.cpp (hunk 1 above)
Comment (ADiea) [inline Sming/SmingCore/Network/TcpConnection.cpp]: I have rebased from l;atest dev and somehow it still sees this as modified although the modified version is already in develop

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 6]
```diff
@@ -99,10 +100,11 @@ void WebsocketClient::onError(err_t err)
 
 bool WebsocketClient::_verifyKey(char* buf, int size)
 {
-	String dd = String(buf);
-	uint8_t s = dd.indexOf("Sec-WebSocket-Accept: ");
+	String keyHeader("Sec-WebSocket-Accept: ");
+	String dd(buf);
```
Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: This code assumes that the header will be in one tcp packet, which is wrong. It may cause broken validation.

```
switch (_mode)
		{
		case wsMode::Connecting:
			if (_verifyKey((char*)data, size) == true)
```

Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: Coding Stype: Replace `dd` with `headerValue`, `s` with `startPos`, `t` with `endPos`.

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 7]
```diff
@@ -99,10 +100,11 @@ void WebsocketClient::onError(err_t err)
...
+	uint8_t s = dd.indexOfIgnoreCase(keyHeader);
```
Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: `s` and `t` can have -1 as value. Replace the type from uint8_t to int.

File: Sming/SmingCore/Network/WebsocketClient.cpp [hunk 8]
```diff
@@ -99,10 +100,11 @@ void WebsocketClient::onError(err_t err)
...
 	uint8_t t = dd.indexOf("\r\n", s);
```
Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: For `s` and `t` - add checks if the return value is `-1` and return false in that case.

File: Sming/SmingCore/Network/WebsocketClient.cpp (hunk 7 above)
Comment (slaff) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: Don't add new method(s) for something that can be achieved with the existing ones. For example you can do something like this:

```
bool WebsocketClient::_verifyKey(char* buf, int size)
{
	String keyHeader("sec-websocket-accept: ");
        String headerValue(buf);
	String headerValueLC = headerValue;
	headerValueLC.toLowerCase();
	int startPos = headerValueLC.indexOf(keyHeader);
	if(startPos == -1) {
		return false;
	}

        // ...
```

File: Sming/Wiring/WString.h [hunk 9]
```diff
@@ -222,6 +222,8 @@ class String
     int indexOf(char ch, unsigned int fromIndex) const;
     int IRAM_ATTR indexOf(const String &str) const;
     int indexOf(const String &str, unsigned int fromIndex) const;
+    int IRAM_ATTR indexOfIgnoreCase(const String &s2) const;
```
Comment (slaff) [inline Sming/Wiring/WString.h]: Same as last comment.

File: Sming/SmingCore/Network/WebsocketClient.cpp (hunk 7 above)
Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: Please keep in mind buf could be 1500bytes(MTU) packet You will create a new string  = heap + make tolower on it. I think this is not optimal that is why I created new search function to ignore case

Comment (ADiea) [inline Sming/SmingCore/Network/WebsocketClient.cpp]: This is why in the first version I prefered to work directly with char* and not involve string copy constructor. We are playing with 1.5KB of heap alloc which you know it's not always available. I think I will revert to first solution of working directly on the buffer.

File: Sming/system/include/stringutil.h [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ *  Contains utility functions for working with char strings.
+ *
+ *  Created on: 26.01.2017
+ *  Author: (github.com/)ADiea
+ */
+
+#ifndef INCLUDE_STRINGUTIL_H_
+#define INCLUDE_STRINGUTIL_H_
+
```
Comment (slaff) [inline Sming/system/include/stringutil.h]: @ADiea Can you add also C linkage so that these functions are also available in C code?

Comment (ADiea) [inline Sming/system/include/stringutil.h]: done

**Comment by ADiea** [comment]
strstri unit test is [here](http://ideone.com/drrMPh)

not all squashes were sucessfull sorry for the big number of commits

**Comment by slaff** [comment]
@ADiea  Thanks for your efforts and for the unit test. Keep up the good work!
